tools/clang/include/clang/SPIRV/ModuleBuilder.h

```diff
@@ -68,7 +68,9 @@ class ModuleBuilder {
 
   /// \brief Creates a SPIR-V basic block. On success, returns the <label-id>
   /// for the basic block. On failure, returns zero.
-  uint32_t createBasicBlock(llvm::StringRef name = "");
+  /// All basic blocks are reachable by default. If isReachable is set to false,
```

Comment:
I just removed the 'reachable' part of code, and then realized we do need it :)

```cpp
while(true) {
  StmtA
  continue;
  StmtB   // This will be inside an unreachable BB. We can choose not to give it a name.
  if(cond) { StmtC } else { StmtD }  // All basic blocks associated with this are unreachable. Its basic blocks are named "if.then", "if.else", and "if.merge"
  for( .... ) { .... }   // All basic blocks associated with this are unreachable. Its basic blocks are named "for.check", "for.body", "for.continue", and "for.merge".
}
```
The code in `doContinueStmt` creates the very first unreachable basic block. StmtB is translated inside that block. But then, that unreachable BB branches to other basic blocks for the IfStmt and ForStmt, and we don't have control over the names of those blocks.

So what happens is that the basic blocks for these gets created with a name and OpName is emitted, but since it is not reachable, the BB itself is not emitted. Therefore validation fails for undefined forward reference to the ID.

The code I've added in `ModuleBuilder::createBasicBlock` ensures that a basic block created from within an unreachableBB is also marked unreachable.

The current implementation adds a boolean to the BasicBlock object so it knows whether it is reachable or not. Alternative solution would be for `createBasicBlock` function to figure out if the current insert point has no name. To do so, it will have to search through the debugNames vector for the result-id of the current insert point and then see if its name is an empty string. This would be much more expensive.